### PR TITLE
Py_trees - Qualification Task Behavior Tree

### PR DESCRIPTION
**Pull request by mpfmorawski**

### CODE REVIEW

**Review by niemiaszek — COMMENTED** [review]
Looks like huge PR (~4k!), I think some cleanup is needed before reviewing.

1.  **Files/classes naming convention**. We should follow PEP, so for files `all_lowercase` and for classes `CapitalWords`. This also applies to changes made by @miki87278 - js files should stay in js convention, but python ones should be changed.
2. **Formatting**. I strongly suggest usage of _Black_ auto-formatter in KNR-AUV organization. I will prepare docs on setup  of _Black_, _isort_ and _Pylint_ in VSC, so you could format whole directory and any further changes with on-save action. Usage of the same formatter by everyone in organization greatly improves readability and ensures that everyone formats our codebase in the same way.
3. **Whole `py_trees_demos` included**.  After brief inspection I see that you don't import any code from these samples in your contributed files. I think it could be better to add note in readme about these demos, so someone could know where to look for those examples, but including them in our repository seems unnecessary.

One final note is to try to force push  to update this branch after #4 merge, so we could have clean PR and repo history.

**Comment by mpfmorawski** [comment]
@niemiaszek 
Suggested changes have been implemented. PR ready for review.

File: okon_actions.py [hunk 1]
```diff
@@ -0,0 +1,390 @@
...
+    def __init__(self,
+                 name: str = "set depth",
+                 okon: Okon = None,
+                 depth: float = 0.6,
+                 delta: float = .05):
+        super().__init__(name)
+        self.okon = okon
+        self.depth = depth
+        self.delta = delta
+        self.logger.debug("%s.__init__()" % (self.__class__.__name__))
```
Comment (niemiaszek) [inline okon_actions.py]: Think about switching to f-strings in all formated strings. Check if they look better than this formatting style

File: okon_actions.py [hunk 2]
```diff
@@ -0,0 +1,390 @@
...
+
+    def initialise(self):
+        self.logger.debug("%s.initialise()" % (self.__class__.__name__))
+
+    def update(self) -> py_trees.common.Status:
+        self.okon.set_depth(self.depth)
+        new_status = py_trees.common.Status.SUCCESS if self.okon.reachedTargetDepth(
```
Comment (niemiaszek) [inline okon_actions.py]: if you use `py_trees.common.Status.SUCCESS` often maybe consider making something like "python typedef", which could short this long name

Comment (niemiaszek) [reply okon_actions.py]: Same for the other py_trees stuff like `FAILURE`, `READ` and `WRITE`

File: okon_actions.py [hunk 3]
```diff
@@ -0,0 +1,390 @@
...
+        self.blackboard.register_key(
+            key="detection", access=py_trees.common.Access.READ)
+        self.logger.debug("%s.__init__()" % (self.__class__.__name__))
+
+    def initialise(self):
+        self.logger.debug("%s.initialise()" % (self.__class__.__name__))
+
+    def update(self):
+        gate = self.blackboard.detection[0]
+        new_status = py_trees.common.Status.SUCCESS if gate['distance'] > self.max_distance else py_trees.common.Status.FAILURE
```
Comment (niemiaszek) [inline okon_actions.py]: Great use of ternary operator 👍

File: okon_actions_example.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+oc = OkonClient(ip="127.0.0.1", port=44210, sync_interval=.05, debug=False)
+oc.connect()
+time.sleep(1.)
+
+def create_root():
+    """ description
+
+    Long description
+    """
```
Comment (niemiaszek) [inline okon_actions_example.py]: I think we left this comment when I described docstrings to You. Consider changing it 😄

File: okon_actions_example.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    rotate_action_2 = Rotate(name = "Turn left",okon=oc.okon, add_angle=-45., delta=1.)
+    set_depth_action_3 = SetDepth(name = "Set Depth to 0.8 m",okon=oc.okon, depth=0.8, delta=0.005)
+    rotate_action_4 = Rotate(name = "Turn right",okon=oc.okon, add_angle=45., delta=1.)
+    root.add_children([set_depth_action_1, rotate_action_2, set_depth_action_3, rotate_action_4])
+    return root
+
+
+##############################################################################
+# Main
+##############################################################################
```
Comment (niemiaszek) [inline okon_actions_example.py]: tbh im not a big fan of such sections. I can see `def main()` right below, so imo no need to add extra block of comment

Comment (niemiaszek) [reply okon_actions_example.py]: Same with `# Execute` section. You could obviously leave single comments to better describe what is happening, but maybe drop whole 2 lines of `#`

File: qualification_task_behavior_tree.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
+""" Script with implementation of behavior tree designed for qualification task (yellow gate)
+
+"""
```
Comment (niemiaszek) [inline qualification_task_behavior_tree.py]: Not sure if you have to add blank line to docstring if you don't add description. Look it up in PEP and just try if it works with single line docstring in VSC hinting

File: qualification_task_behavior_tree.py [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+
+    sequence_2.add_children([
+                    wait_for_2_secs,
+                    stop_okon,
+                    wait_for_a_while,
+                    exit_action])
+
+    root.add_children([
+                    sequence_1,
+                    sequence_2])
```
Comment (niemiaszek) [inline qualification_task_behavior_tree.py]: This looks really weirdy formatted to me with those brackets `[` hanging alone. Was it formatted with _Black_?

File: qualification_task_behavior_tree.py [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+    root = create_root()
+
+####################
+# Execute
+####################
+
+    root.setup_with_descendants()
+    for i in range(0, 200):
+        try:
+            print(f"\n--------- Tick {i} ---------\n")
```
Comment (niemiaszek) [inline qualification_task_behavior_tree.py]: Look up character filling in f-strings. Maybe you wouldn't need to type manually all these `-`

**Review by niemiaszek — COMMENTED** [review]
Very nice improvement from initial version of this PR. No issues in logic, it does what it should. Just some styling tips

File: okon_actions_example.py [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+    """ Main function of the script"""
+    py_trees.logging.level = py_trees.logging.Level.DEBUG
+
+    root = create_root()
+    root.setup_with_descendants()
+    
+    for i in range(0, 30):
+        try:
+            tick_string = "Tick"
+            print(f"\n{tick_string:.>25} {i :.<25}\n")
```
Comment (niemiaszek) [inline okon_actions_example.py]: Today I learnt about f-string nesting, symetrical filling and debugging 😄. Check this out 🤓 
```suggestion
    for tick in range(0, 30):
        try:
            print(f"\n{f'{tick = }':^25}\n")

**Comment by mpfmorawski** [comment]
@niemiaszek
Suggested changes have been implemented. PR ready for review.

File: qualification_task_behavior_tree.py [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+    return root
+
+
+def main():
+    py_trees.logging.level = py_trees.logging.Level.DEBUG
+
+    root = create_root()
+
+    root.setup_with_descendants()
+    for tick in range(0, 200):
```
Comment (niemiaszek) [inline qualification_task_behavior_tree.py]: no need to add 0 as starting point in range, as it is default parameter

File: okon_actions_example.py [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+
+
+def main():
+    """Main function of the script"""
+    py_trees.logging.level = py_trees.logging.Level.DEBUG
+
+    root = create_root()
+    root.setup_with_descendants()
+
+    for tick in range(0, 30):
```
Comment (niemiaszek) [inline okon_actions_example.py]: same as in comment above. Must have missed this

**Review by niemiaszek — COMMENTED** [review]
Minor issues

**Comment by mpfmorawski** [comment]
@niemiaszek
Minor changes implemented

**Review by niemiaszek — APPROVED** [review]
LGTM :rocket:. Thanks for so much work put into polishing these changes :+1: 

Before you merge, look into all merging options:
- Merge 38 commits
- Squash and merge (1 commit made out of all changes)
- Rebase and merge

Try to understand the difference between them. In my opinion, for 38 commits in here we won't need any intermediate state in our main branch - only final status of this PR matters to us. Therfore, IMO the best option is squash and merge.

**Pull request merged**